### ISSUE TITLE
Support for AES-128-GCM decryption

### ISSUE THREAD

**Issue by flyingolivier** [onelogin/ruby-saml#541]
We are trying to use `ruby-saml` to implement a Service Provider (SP) that talks to a Shibboleth Identity Provider (IDP). The IDP returns an encrypted assertion using the `AES-128-GCM` algorithm. `ruby-saml` is failing to decrypt the assertion and returns the following error message:

> NoMethodError (undefined method `[]' for nil:NilClass)

Reading [the code](https://github.com/onelogin/ruby-saml/blob/master/lib/onelogin/ruby-saml/utils.rb#L251-L258 ) it appears that CBC mode algoritms such as `DES-EDE3-CBC`, `AES-128-CBC`, `AES-192-CBC` and `AES-256-CBC` are suported by `ruby-saml` but the `AES-GCM` modes are not.

Shibboleth recently [changed its encryption defaults to favor AES-GCM for XML encryption](https://wiki.shibboleth.net/confluence/display/IDP4/GCMEncryption). Considering that Shibboleth is widely used, It is likely that more and more Service Providers will be facing this issue.

Further investigations show that the OpenSSL library used by this gem already supports CGM modes.

```ruby
 irb(main):065:0> OpenSSL::Cipher.ciphers.include?("aes-128-gcm")
=> true
```

My knowledge in encryption/decryption algorithms is limited but I will gladly provide more information to see this support added.

**Comment by ahukkanen** [comment]
Here is the method in `ruby-saml` that does the decryption of the encrypted SAML message:
https://github.com/onelogin/ruby-saml/blob/7423f03618c0f78f072d31132375cb4e2d89b282/lib/onelogin/ruby-saml/utils.rb#L250-L273

Here is how you can modify that to support AES-128-GCM, AES-192-GCM and AES-256-GCM:

```ruby
OneLogin::RubySaml::Utils.class_eval do
  def self.retrieve_plaintext(cipher_text, symmetric_key, algorithm)
    case algorithm
      when 'http://www.w3.org/2001/04/xmlenc#tripledes-cbc' then cipher = OpenSSL::Cipher.new('DES-EDE3-CBC').decrypt
      when 'http://www.w3.org/2001/04/xmlenc#aes128-cbc' then cipher = OpenSSL::Cipher.new('AES-128-CBC').decrypt
      when 'http://www.w3.org/2001/04/xmlenc#aes192-cbc' then cipher = OpenSSL::Cipher.new('AES-192-CBC').decrypt
      when 'http://www.w3.org/2001/04/xmlenc#aes256-cbc' then cipher = OpenSSL::Cipher.new('AES-256-CBC').decrypt
      when 'http://www.w3.org/2001/04/xmlenc#rsa-1_5' then rsa = symmetric_key
      # ADDED =>
      when 'http://www.w3.org/2001/04/xmlenc#rsa-oaep-mgf1p' then oaep = symmetric_key
      when 'http://www.w3.org/2009/xmlenc11#aes128-gcm' then auth_cipher = OpenSSL::Cipher.new('AES-128-GCM').decrypt
      when 'http://www.w3.org/2009/xmlenc11#aes192-gcm' then auth_cipher = OpenSSL::Cipher.new('AES-192-GCM').decrypt
      when 'http://www.w3.org/2009/xmlenc11#aes256-gcm' then auth_cipher = OpenSSL::Cipher.new('AES-256-GCM').decrypt
      # <= ADDED
    end

    if cipher
      iv_len = cipher.iv_len
      data = cipher_text[iv_len..-1]
      cipher.padding, cipher.key, cipher.iv = 0, symmetric_key, cipher_text[0..iv_len-1]
      assertion_plaintext = cipher.update(data)
      assertion_plaintext << cipher.final
    # ADDED =>
    elsif auth_cipher
      iv_len, text_len, tag_len = auth_cipher.iv_len, cipher_text.length, 16
      data = cipher_text[iv_len..text_len-1-tag_len]
      auth_cipher.padding = 0
      auth_cipher.key = symmetric_key
      auth_cipher.iv = cipher_text[0..iv_len-1]
      auth_cipher.auth_data = ''
      auth_cipher.auth_tag = cipher_text[text_len-tag_len..-1]
      assertion_plaintext = auth_cipher.update(data)
      assertion_plaintext << auth_cipher.final
    # <= ADDED
    elsif rsa
      rsa.private_decrypt(cipher_text)
    elsif oaep
      oaep.private_decrypt(cipher_text, OpenSSL::PKey::RSA::PKCS1_OAEP_PADDING)
    else
      cipher_text
    end
  end
end
```

There is a difference with the GCM ciphres as with those, there is an extra "authentication tag" transferred with the ciphertext as the 16 last bytes of that text. This is explained here:
https://tools.ietf.org/html/rfc5116#section-5.1

> An authentication tag with a length of 16 octets (128
> bits) is used.  The AEAD_AES_128_GCM ciphertext is formed by
> appending the authentication tag provided as an output to the GCM
> encryption operation to the ciphertext that is output by that
> operation.

Therefore, you need to extract the 16 last bytes of the message and define it as the cipher's `auth_tag`.

**Comment by silppuri** [comment]
Hi, we also stumbled upon this today. Is there any progress with the issue? We would be happy to help by contributing with a fix to this. In the meantime, we can use a fork of the library but would like to see this supported by the upstream.

**Comment by sergeyki** [comment]
Hi, I also faced the issue with aes-128-gcm today. Any updates?

**Comment by pitbulk** [comment]
PR with unitest is welcome

**Comment by silppuri** [comment]
Hi, just an update that I started working on this. Hopefully I'll get the PR for review early next week

**Comment by pitbulk** [comment]
Nice, thanks!

**Comment by silppuri** [comment]
Hi @pitbulk the changes are implemented here https://github.com/onelogin/ruby-saml/pull/557

**Comment by pitbulk** [comment]
Merged on 1.12 branch
